relays/ethereum/src/message_race_loop.rs

```diff
@@ -0,0 +1,337 @@
+use futures::{
+	future::FutureExt,
+	stream::{FusedStream, StreamExt},
+};
+use std::{
+	fmt::Debug,
+	ops::RangeInclusive,
+	time::{Duration, Instant},
+};
+
+/// One of races within lane.
+pub trait MessageRace {
+	/// Header id of the race source.
+	type SourceHeaderId: Debug + Clone + PartialEq;
+	/// Header id of the race source.
+	type TargetHeaderId: Debug + Clone + PartialEq;
+
+	/// Message nonce used in the race.
+	type MessageNonce: Debug + Clone;
+	/// Proof that is generated and delivered in this race.
+	type Proof: Clone;
+
+	/// Name of the race source.
+	fn source_name() -> String;
+	/// Name of the race target.
+	fn target_name() -> String;
+}
+
+/// State of race source client.
+type SourceClientState<P> = ClientState<<P as MessageRace>::SourceHeaderId, <P as MessageRace>::TargetHeaderId>;
+
+/// State of race target client.
+type TargetClientState<P> = ClientState<<P as MessageRace>::TargetHeaderId, <P as MessageRace>::SourceHeaderId>;
+
+/// One of message lane clients, which is source client for the race.
+#[async_trait(?Send)]
+pub trait SourceClient<P: MessageRace> {
+	/// Type of error this clients returns.
+	type Error: std::fmt::Debug + MaybeConnectionError;
+
+	/// Return latest nonce that is known to the source client.
+	async fn latest_nonce(
+		&self,
+		at_block: P::SourceHeaderId,
+	) -> Result<(P::SourceHeaderId, P::MessageNonce), Self::Error>;
+	/// Generate proof for delivering to the target client.
+	async fn generate_proof(
+		&self,
+		at_block: P::SourceHeaderId,
+		nonces: RangeInclusive<P::MessageNonce>,
+	) -> Result<(P::SourceHeaderId, RangeInclusive<P::MessageNonce>, P::Proof), Self::Error>;
+}
+
+/// One of message lane clients, which is target client for the race.
+#[async_trait(?Send)]
+pub trait TargetClient<P: MessageRace> {
+	/// Type of error this clients returns.
+	type Error: std::fmt::Debug + MaybeConnectionError;
+
+	/// Return latest nonce that is known to the target client.
+	async fn latest_nonce(
+		&self,
+		at_block: P::TargetHeaderId,
+	) -> Result<(P::TargetHeaderId, P::MessageNonce), Self::Error>;
+	/// Submit proof to the target client.
+	async fn submit_proof(
+		&self,
+		generated_at_block: P::SourceHeaderId,
+		nonces: RangeInclusive<P::MessageNonce>,
+		proof: P::Proof,
+	) -> Result<RangeInclusive<P::MessageNonce>, Self::Error>;
+}
+
+/// Race strategy.
+pub trait RaceStrategy<SourceHeaderId, TargetHeaderId, MessageNonce, Proof> {
+	/// Should return true if nothing has to be synced.
+	fn is_empty(&self) -> bool;
+	/// Called when latest nonce is updated at source node of the race.
+	fn source_nonce_updated(&mut self, at_block: SourceHeaderId, nonce: MessageNonce);
+	/// Called when latest nonce is updated at target node of the race.
+	fn target_nonce_updated(
+		&mut self,
+		nonce: MessageNonce,
+		race_state: &mut RaceState<SourceHeaderId, TargetHeaderId, MessageNonce, Proof>,
+	);
+	/// Should return `Some(_)` with if we need to deliver proof of given nonces from
```

Comment:
I think the docs needs an update here, it returns `Some` with the range of nonces that require proof, right?